File: ml/rook/observation.py
"""AlphaRook's window into the game — port of src/lib/alpharook/observation.ts.

The rule-integrity contract: an Observation for a seat contains exactly what
a human in that chair knows — their own cards, everything played face-up, the
bidding, and the go-down only if they took the bid. The neural encoder
consumes an Observation, never a Game, so cheating is impossible by
construction; tests/test_leak.py proves hidden cards can't influence it.
"""

from __future__ import annotations

from dataclasses import dataclass, field

from .cards import SEATS, suit_of
from .engine import Game


@dataclass
class Observation:
    seat: int
    phase: int
    turn: int | None
    hand_number: int
    dealer: int | None

    # auction (fully public)
    bids: list[int | None]          # per seat: None silent, PASS, or value
    high_bid: int | None
    bid_winner: int | None
    trump: int | None

    # my private cards
    hand: list[int]
    my_go_down: list[int] | None    # only if I won the bid

    # table state (fully public)
    trick_leader: int | None
    trick_plays: list[tuple[int, int]]
    completed_tricks: list[tuple[int, list[tuple[int, int]], int, int]]
    tricks_won: list[int]
    points_taken: list[int]
    scores: list[int]


def observe(g: Game, seat: int) -> Observation:
    return Observation(
        seat=seat,
        phase=g.phase,
        turn=g.turn,
        hand_number=g.hand_number,
        dealer=g.dealer,
        bids=list(g.bids),
        high_bid=g.high_bid,
        bid_winner=g.bid_winner,
        trump=g.trump,
        hand=list(g.hands[seat]),
        my_go_down=list(g.go_down) if g.bid_winner == seat else None,
        trick_leader=g.trick_leader,
        trick_plays=list(g.trick_plays),
        completed_tricks=list(g.completed_tricks),
        tricks_won=list(g.tricks_won),
        points_taken=list(g.points_taken),
        scores=list(g.scores),
    )


def seen_cards(o: Observation) -> list[int]:
    out = [c for _, plays, _, _ in o.completed_tricks for _, c in plays]
    out.extend(c for _, c in o.trick_plays)
    out.extend(o.hand)
    if o.my_go_down:
        out.extend(o.my_go_down)
    return out


def unseen_cards(o: Observation) -> list[int]:
    seen = set(seen_cards(o))
    return [c for c in range(40) if c not in seen]


def known_voids(o: Observation) -> list[set[int]]:
    """Per seat, the suits they've shown they're out of."""
    voids: list[set[int]] = [set(), set(), set(), set()]
    all_plays = [plays for _, plays, _, _ in o.completed_tricks]
    if o.trick_plays:
        all_plays.append(o.trick_plays)
    for plays in all_plays:
        if not plays:
            continue
        lead = suit_of(plays[0][1])
        for s, c in plays:
            if suit_of(c) != lead:
                voids[s].add(lead)
    return voids


def hand_sizes(o: Observation) -> list[int]:
    played_this_trick = {s for s, _ in o.trick_plays}
    base = 9 - len(o.completed_tricks)
    return [base - (1 if s in played_this_trick else 0) for s in SEATS]

File: src/lib/alpharook/observation.ts
// AlphaRook's window into the game. This is the ONLY thing an agent may see.
//
// The rule-integrity contract: an Observation for a seat contains exactly the
// information a human in that chair would have — their own cards, everything
// played face-up, the bidding, and (only if they took the bid) the widow/
// go-down they themselves handled. Never other players' hands, never a widow
// they didn't win. Agents receive an Observation, not a GameDoc, so cheating
// is impossible by construction; observation.test.ts additionally proves no
// hidden card can leak through.

import { GameDoc, Card, Seat, Suit, Team, TrickRecord, Phase, SUITS } from '../game/types';

export interface Observation {
    seat: Seat;
    phase: Phase;
    turn: Seat | null;
    handNumber: number;
    dealer: Seat | null;

    // auction (fully public)
    bids: Partial<Record<Seat, number | 'pass'>>;
    highBid: number | null;
    bidWinner: Seat | null;
    trump: Suit | null;

    // my private cards
    hand: Card[];
    /** the go-down I chose — null unless I am the bid winner */
    myGoDown: Card[] | null;

    // table state (fully public)
    trickLeader: Seat | null;
    trickPlays: { seat: Seat; card: Card }[];
    completedTricks: TrickRecord[];
    tricksWon: Record<Team, number>;
    pointsTaken: Record<Team, number>;
    scores: Record<Team, number>;
}

/** Build what `seat` can legally see. Pure; never exposes hidden zones. */
export const observe = (g: GameDoc, seat: Seat): Observation => ({
    seat,
    phase: g.phase,
    turn: g.turn,
    handNumber: g.handNumber,
    dealer: g.dealer,
    bids: { ...g.bids },
    highBid: g.highBid,
    bidWinner: g.bidWinner,
    trump: g.trump,
    hand: g.hands[seat].map((c) => ({ ...c })),
    myGoDown: g.bidWinner === seat ? g.goDown.map((c) => ({ ...c })) : null,
    trickLeader: g.trickLeader,
    trickPlays: g.trickPlays.map((p) => ({ seat: p.seat, card: { ...p.card } })),
    completedTricks: g.completedTricks.map((t) => ({
        leader: t.leader,
        plays: t.plays.map((p) => ({ seat: p.seat, card: { ...p.card } })),
        winner: t.winner,
        points: t.points,
    })),
    tricksWon: { ...g.tricksWon },
    pointsTaken: { ...g.pointsTaken },
    scores: { ...g.scores },
});

// ---------------------------------------------------------------------------
// Derived knowledge — the "memory" a careful human keeps at the table.
// ---------------------------------------------------------------------------

/** Every card the observer has seen face-up or holds. */
export const seenCards = (o: Observation): Card[] => [
    ...o.completedTricks.flatMap((t) => t.plays.map((p) => p.card)),
    ...o.trickPlays.map((p) => p.card),
    ...o.hand,
    ...(o.myGoDown ?? []),
];

/** Cards whose location the observer does not know. */
export const unseenCards = (o: Observation): Card[] => {
    const seen = new Set(seenCards(o).map((c) => `${c.suit}-${c.number}`));
    const out: Card[] = [];
    for (const suit of SUITS) {
        for (let n = 5; n <= 14; n++) {
            if (!seen.has(`${suit}-${n}`)) out.push({ suit, number: n });
        }
    }
    return out;
};

/** Suits each opponent/partner has shown they are out of (failed to follow). */
export const knownVoids = (o: Observation): Record<Seat, Set<Suit>> => {
    const voids: Record<Seat, Set<Suit>> = {
        A1: new Set(), B1: new Set(), A2: new Set(), B2: new Set(),
    };
    const allTricks = [
        ...o.completedTricks.map((t) => t.plays),
        ...(o.trickPlays.length > 0 ? [o.trickPlays] : []),
    ];
    for (const plays of allTricks) {
        if (plays.length === 0) continue;
        const lead = plays[0].card.suit;
        for (const p of plays) {
            if (p.card.suit !== lead) voids[p.seat].add(lead);
        }
    }
    return voids;
};

/** How many cards each seat is currently holding. */
export const handSizes = (o: Observation): Record<Seat, number> => {
    const playedThisTrick = new Set(o.trickPlays.map((p) => p.seat));
    const base = 9 - o.completedTricks.length;
    return {
        A1: base - (playedThisTrick.has('A1') ? 1 : 0),
        B1: base - (playedThisTrick.has('B1') ? 1 : 0),
        A2: base - (playedThisTrick.has('A2') ? 1 : 0),
        B2: base - (playedThisTrick.has('B2') ? 1 : 0),
    };
};
